_applications/equoid.adoc

```diff
@@ -0,0 +1,219 @@
++
+....
+oc new-app --template=infinispan-ephemeral \
+    -l app=datagrid \
+    -p APPLICATION_NAME=datagrid \
+    -p NAMESPACE=`oc project -q` \
+    -p APPLICATION_USER=<YOUR_APP_USERNAME> \
+    -p APPLICATION_PASSWORD=<YOUR_APP_PASSWORD> \
+    -p MANAGEMENT_USER=<YOUR_MGMT_USERNAME> \
+    -p MANAGEMENT_PASSWORD=<YOUR_MGMT_PASSWORD>
+....
++
+
+. Launch Data Publisher
++
+We can next begin publishing our streaming data by starting an instance of the data publisher. `OP_MODE` described later should be set to the mode you wish. Also, `DATA_URL_PRIMARY` should be set to the location of your data set or data potential values file. 
++
+....
+oc new-app \
+    -l app=publisher \
+    -e OP_MODE={single|dual|linear} \
+    -e DATA_URL_PRIMARY=<YOUR_DATASET_URL> \
+    --image-stream=`oc project -q`/fuse-java \
+    https://github.com/radanalyticsio/equoid-data-publisher
+....
+
+. Launch Data Handler
++
+We are now ready to start our data handler. Note that `INFINISPAN_HOST` and `INFINISPAN_PORT` need to be set to appropriate values, in the case of the template provided, these are `datagrid-hotrod` and `11222` respectively. In addition, `WINDOW_SECONDS` is the number of second for which you wish to track the top-k items, `SLIDE_SECONDS` represents the number of seconds to slide your data window each iteration, and `BATCH_SECONDS` is the size of samples you wish to acquire from the message stream at a time. Finally, `OP_MODE` needs to be set to the same value as set for the data publisher.
++
+....
+oc new-app --template=oshinko-scala-spark-build-dc \
+    -l app=handler \
+    -p SBT_ARGS=assembly \
+    -p APPLICATION_NAME=equoid-data-handler \
+    -p GIT_URI=https://github.com/radanalyticsio/equoid-data-handler \
+    -p APP_MAIN_CLASS=io.radanalytics.equoid.DataHandler \
+    -e INFINISPAN_HOST=<YOUR_HOSTNAME> \
+    -e INFINISPAN_PORT=<YOUR_PORT> \
+    -e WINDOW_SECONDS=<YOUR_WINDOW_SIZE> \
+    -e SLIDE_SECONDS=<YOUR_SLIDING_SIZE> \
+    -e BATCH_SECONDS=<YOUR_BATCH_SIZE> \
+    -e OP_MODE={single|dual|linear} \
+    -p SPARK_OPTIONS='--driver-java-options=-Dvertx.cacheDirBase=/tmp'
+....
+
+. Launch Web UI and Set Up Keycloak
++
+Equoid's web UI is launched using a script developed by one of Equoid's contributors, Jiri Kremser. The following calls will utilize that script to set up the UI as well as a Keycloak instance for users and login. Finally, an edit role is given to the project's service account so additional data handler instances can be created/modified. 
++
+....
+BASE_URL="https://raw.githubusercontent.com/radanalyticsio/equoid-ui/master/ocp/"
+curl -sSL $BASE_URL/ocp-apply.sh | \
+    BASE_URL="$BASE_URL" \
+    KC_REALM_PATH="web-ui/keycloak/realm-config" \
+    bash -s stable
+oc policy add-role-to-user edit system:serviceaccount:`oc project -q`:default
+....
```

Comment:
this part makes me really nervous, i hate telling users to curl a random script from the internet and run it. i don't have a suggestion to make it better, but i just needed to say something.

given where we are at, i'm not sure i see a clear path to making this better but i think it highlights a sticky infrastructure issue with deploying keycloak.

Reply:
ACK

Reply:
The main reason for this was not to repeat the same code in multiple repositories. What about doing the sha1 check, would it help? Something like:

```bash
curl -fLs https://raw.githubusercontent.com/eldritchjs/equoid-ui/master/ocp/ocp-apply.sh | tee ocp-apply.sh | echo "da39a3ee5e6b4b0d3255bfef95601890afd80709  ocp-apply.sh" | sha1sum -c - || exit 1
```

`tee` will split the output to file and stdout and if the sha1 check fails, it will do the `exit 1`